src/main/cpp/src/hyper_log_log_plus_plus.cu

```diff
@@ -0,0 +1,958 @@
+      uint64_t const hash = static_cast<uint64_t>(hashs.element<int64_t>(i));
+      reg_idx             = hash >> idx_shift;
+      // get the leading zeros
+      reg_v = static_cast<int>(cuda::std::countl_zero((hash << precision) | w_padding) + 1ULL);
+    }
+
+    cuda::atomic_ref<int32_t, cuda::thread_scope_block> register_ref(shared_data[reg_idx]);
+    register_ref.fetch_max(reg_v, cuda::memory_order_relaxed);
+  }
+  __syncthreads();
+
+  // compact from register values (int array) to long array
+  // each long holds 10 integers, note reg value < 64 which means the bits from
+  // 7 to highest are all 0.
+  if (tid * REGISTERS_PER_LONG < num_registers_per_sketch) {
+    int start = tid * REGISTERS_PER_LONG;
+    int end   = (tid + 1) * REGISTERS_PER_LONG;
+    if (end > num_registers_per_sketch) { end = num_registers_per_sketch; }
+
+    int64_t ret = 0;
+    for (int i = 0; i < end - start; i++) {
+      int shift   = i * REGISTER_VALUE_BITS;
+      int64_t reg = shared_data[start + i];
+      ret |= (reg << shift);
+    }
+
+    output[tid][0] = ret;
+  }
+}
+
+std::unique_ptr<cudf::scalar> reduce_hllpp(cudf::column_view const& input,
+                                           int64_t const precision,
+                                           rmm::cuda_stream_view stream,
+                                           rmm::device_async_resource_ref mr)
+{
+  int64_t num_registers_per_sketch = 1L << precision;
+  // 1. compute all the hashs
+  auto input_table_view = cudf::table_view{{input}};
+  auto hash_col         = xxhash64(input_table_view, SEED, stream, mr);
+  hash_col->set_null_mask(cudf::detail::copy_bitmask(input, stream, mr), input.null_count());
+  auto d_hashs = cudf::column_device_view::create(hash_col->view(), stream);
+
+  // 2. generate long columns, the size of each long column is 1
+  auto num_long_cols      = num_registers_per_sketch / REGISTERS_PER_LONG + 1;
+  auto const results_iter = cudf::detail::make_counting_transform_iterator(0, [&](int i) {
+    return cudf::make_numeric_column(cudf::data_type{cudf::type_id::INT64},
+                                     1 /**num_groups*/,
+                                     cudf::mask_state::UNALLOCATED,
+                                     stream,
+                                     mr);
+  });
+  auto children =
+    std::vector<std::unique_ptr<cudf::column>>(results_iter, results_iter + num_long_cols);
+  auto d_results = [&] {
+    auto host_results_pointer_iter =
+      thrust::make_transform_iterator(children.begin(), [](auto const& results_column) {
+        return results_column->mutable_view().template data<int64_t>();
+      });
+    auto host_results_pointers =
+      std::vector<int64_t*>(host_results_pointer_iter, host_results_pointer_iter + children.size());
+    return cudf::detail::make_device_uvector_async(
+      host_results_pointers, stream, cudf::get_current_device_resource_ref());
+  }();
```

Comment:
`host_results_pointers` can be destroyed before data is copied to device. Since we don't like stream sync, just move it out of the code block:
```suggestion
    auto host_results_pointer_iter =
      thrust::make_transform_iterator(children.begin(), [](auto const& results_column) {
        return results_column->mutable_view().template data<int64_t>();
      });
    auto host_results_pointers =
      std::vector<int64_t*>(host_results_pointer_iter, host_results_pointer_iter + children.size());
  auto d_results = cudf::detail::make_device_uvector_async(
      host_results_pointers, stream, cudf::get_current_device_resource_ref());
```

Reply:
done